### PR TITLE
[Feature] add multi-table transaction stream load support

### PR DESCRIPTION
**Pull request by meegoo**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

- Add WriteMultipleTablesWithTransaction.java example with TransactionAssembler (KeyedProcessFunction) for Kafka transaction assembly
- Introduce DefaultStarRocksRowData for dynamic multi-table routing (database-name=*, table-name=*)
- Extend StarRocksRowData interface with isTransactionEnd() for txn boundary control
- Use SinkFunctionFactory.createSinkFunction() for standard Flink SinkFunction
- Add per-table StreamLoadTableProperties configuration support
- Add docs (EN/CN) for multi-table transaction stream load usage
- Add MultiTableTransactionITTest and StreamLoadManagerMultiTableTest
- Enhance StreamLoader, TransactionStreamLoader, StreamLoadManagerV2 for multi-table txn
- Replace non-compilable TxnBatch/writeBatch design with Flink DataStream API patterns

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function

### CODE REVIEW

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by meegoo** [comment]
@cursor review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by meegoo** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

File: starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/TransactionTableRegion.java [hunk 1]
```diff
@@ -165,10 +171,17 @@ public LabelGenerator getLabelGenerator() {
...
+        // Guard against injecting a new shared label while a retry is in progress.
+        // numRetries is mutated under this same monitor (see fail()), so the check
+        // and the assignment are now atomic with respect to concurrent fail() calls.
         if (numRetries > 0 && label != null) {
-            return;
+            LOG.warn("[MultiTxn] setLabel called with label={} while numRetries={}, "
+                    + "existing label={}. Rejecting to preserve retry consistency.",
+                    label, numRetries, this.label);
+            throw new IllegalStateException(
+                    "Cannot set label while region is retrying (numRetries=" + numRetries
```
Comment (banmoy) [inline starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/TransactionTableRegion.java]: **[Critical] `setLabel()` behavior change may break non-multi-table retry path**

Original code silently returned when `numRetries > 0 && label != null`; the new code throws `IllegalStateException`. While `ensureSharedTransaction()` checks `isRetrying()` before calling `setLabel()`, the standard `TransactionStreamLoader.begin(region)` does **not** check `isRetrying()`. Under rare timing in the existing non-multi-table retry path, this could surface as a regression.

Suggestion: either add a `multiTableMode` flag to only throw in multi-table mode, or exhaustively verify (and document) that the non-multi-table path can never reach this condition.

File: starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/TransactionTableRegion.java [hunk 2]
```diff
@@ -227,6 +257,61 @@ private void switchChunk() {
         activeChunk = new Chunk(properties.getDataFormat(), chunkIdGenerator.getAndIncrement());
     }
 
+    /**
+     * Called by the task thread when a txnEnd marker is received (multi-table mode).
+     *
+     * <p>Moves the active chunk to the inactive queue so the manager thread can
```
Comment (banmoy) [inline starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/TransactionTableRegion.java]: **[Critical] Misleading Javadoc — also called from manager thread**

The comment says "This runs on the task thread which is single-threaded in Flink, so there is no concurrent write() call." However, the savepoint handler in `DefaultStreamLoadManager` also calls `switchChunkForCommit()` from the **manager thread**:

```java
// savepoint handling in DefaultStreamLoadManager:
for (TransactionTableRegion region : flushQ) {
    region.switchChunkForCommit();
}
```

At savepoint time, the task thread is blocked in `flush()`, so there should be no contention — but this assumption is subtle and fragile. Please update the Javadoc to document both call sites (task thread on txnEnd, manager thread during savepoint).

File: starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/DefaultStreamLoadManager.java [hunk 3]
```diff
@@ -438,7 +937,7 @@ private long calculateNextWaitTime(long currentWaitTime) {
     private void finishFlush() {
         LOG.info("Stream load manager flush finished - currentCacheBytes: {}, maxCacheBytes: {}, allRegionsCommitted: {}",
                 currentCacheBytes.get(), maxCacheBytes, allRegionsCommitted);
-        AssertNotException();
+        checkAndThrowException();
         savepoint = false;
     }
```
Comment (banmoy) [inline starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/DefaultStreamLoadManager.java:1408]: **[Critical] Double `savepoint = false` assignment**

`finishFlush()` sets `savepoint = false` here, and the `finally` block in `flush()` (line ~881) also sets `savepoint = false`. On the happy path the field is always written twice. More importantly, `checkAndThrowException()` can throw, and after the throw the `finally` block correctly clears `savepoint` anyway.

Suggest removing `savepoint = false` from `finishFlush()` since the `finally` block in `flush()` already handles it. This makes the control flow clearer.

File: starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/TransactionTableRegion.java [hunk 4]
```diff
@@ -110,11 +111,16 @@ public TransactionTableRegion(String uniqueKey,
...
         this.maxRetries = maxRetries;
         this.retryIntervalInMs = retryIntervalInMs;
-        initHeaders(properties);
     }
 
     private void initHeaders(StreamLoadTableProperties properties) {
         headers.putAll(properties.getProperties());
+        // Include db and table headers so that transaction stream load
+        // (/api/transaction/load) routes data to the correct table.
+        // In multi-table transaction mode, each region targets a different
```
Comment (banmoy) [inline starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/TransactionTableRegion.java:202]: **[Important] `db` and `table` headers added unconditionally — affects non-multi-table mode too**

`headers.put("db", database)` and `headers.put("table", table)` are now always added, even in non-multi-table mode. For single-table transaction stream load, these extra headers should be benign since StarRocks uses the URL path, but it is a behavioral change for all users.

Consider gating this behind a multi-table mode check, or verifying with StarRocks documentation that these headers are harmless in single-table transaction mode.

File: starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/DefaultStreamLoadManager.java [hunk 5]
```diff
@@ -229,23 +421,314 @@ public void init() {
...
+                ensureSharedTransaction();
+            }
+
+            // Wait for any in-flight autonomous flushes to complete.
+            for (TransactionTableRegion region : regionSnapshot) {
+                if (region.isFlushing() || region.isRetrying()) {
+                    LOG.debug("[MultiTxn] Region {} still flushing/retrying, will retry next scan",
+                            region.getUniqueKey());
+                    return;
+                }
```
Comment (banmoy) [inline starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/DefaultStreamLoadManager.java:801]: **[Important] No timeout for stuck `commitInFlight`**

If a region's HTTP load is perpetually failing and retrying, `processMultiTableCommit()` returns on every scan and `commitInFlight` stays `true` indefinitely. This stalls the entire pipeline with no escape hatch. The savepoint path will eventually catch this via `flushTimeoutMs`, but there's no timeout within `processMultiTableCommit()` itself for the normal (non-savepoint) code path.

Suggest adding a timeout or max iteration count for the commit-in-flight phase. If exceeded, abort the commit cycle, propagate the error, and allow recovery.

File: starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/DefaultStreamLoadManager.java [hunk 6]
```diff
@@ -229,23 +421,314 @@ public void init() {
...
+            txnCoordinator.reset();
+            LOG.info("[MultiTxn] Recycled empty shared transaction (rolled back)");
+        }
+
+        // Clear labels and open a fresh shared transaction.
+        for (TransactionTableRegion region : flushQ) {
+            region.setLabel(null);
+        }
+        ensureSharedTransaction();
     }
```
Comment (banmoy) [inline starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/DefaultStreamLoadManager.java:1165]: **[Important] Implicit ordering dependency in `recycleSharedTransaction()`**

Between clearing labels (line ~727) and re-opening the shared txn via `ensureSharedTransaction()`, if an autonomous flush were triggered, it could start an HTTP load with `label=null`, causing `TransactionStreamLoader.begin()` to generate an independent label — violating atomicity.

This is mitigated because `recycleSharedTransaction()` runs on the manager thread and the caller has a `continue` that skips the flush selection loop for this iteration. But this coupling is implicit and fragile. Please add a comment explaining why the `continue` after the recycle call is critical for correctness.

File: starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/DefaultStreamLoadManager.java [hunk 7]
```diff
@@ -110,6 +121,17 @@ enum State {
 
     private final Queue<TransactionTableRegion> flushQ = new ConcurrentLinkedQueue<>();
 
+    /** Per-partition region index for multi-table transaction mode. */
+    private final Map<Integer, List<TransactionTableRegion>> partitionRegions = new ConcurrentHashMap<>();
```
Comment (banmoy) [inline starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/DefaultStreamLoadManager.java:171]: **[Important] `partitionRegions` map never cleaned up**

Entries are added in `getCacheRegion()` but never removed. For long-running Flink jobs with many distinct partitions over time (e.g., rebalance events), this could lead to unbounded memory growth. Consider cleaning up entries when partitions are idle, or document the expected memory profile.

File: src/main/java/com/starrocks/connector/flink/table/sink/StarRocksSinkOptions.java [hunk 8]
```diff
@@ -506,6 +529,17 @@ private void validateMergeCommit() {
...
+    private void validateMultiTableTransaction() {
+        if (!tableOptions.get(SINK_MULTI_TABLE_TXN_ENABLED)) {
+            return;
+        }
+        if (sinkSemantic == StarRocksSinkSemantic.EXACTLY_ONCE) {
+            throw new ValidationException(
+                    "Multi-table transaction stream load does not support exactly-once semantics. " +
+                    "Please use at-least-once semantics with '" + SINK_MULTI_TABLE_TXN_ENABLED.key() + "'.");
+        }
+    }
```
Comment (banmoy) [inline src/main/java/com/starrocks/connector/flink/table/sink/StarRocksSinkOptions.java:552]: **[Important] Missing `sink.version=V2` validation**

`validateMultiTableTransaction()` checks for exactly-once incompatibility but does **not** enforce that sink version is V2. The doc says "Requires `sink.version=V2`" but if someone enables multi-table mode with V1 sink, the behavior is undefined.

Suggest adding:
```java
if (!SinkVersion.V2.name().equalsIgnoreCase(getSinkVersion())) {
    throw new ValidationException("Multi-table transaction requires sink.version=V2");
}
```

File: starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/TransactionTableRegion.java [hunk 9]
```diff
@@ -198,25 +221,32 @@ public long getAge() {
         return age.get();
     }
 
+    private static final int MAX_SPIN_ATTEMPTS = 10;
+    private static final long SPIN_BACKOFF_NANOS = 1000L; // 1 microsecond initial backoff
+
     @Override
     public int write(byte[] row) {
         if (row == null) {
             return 0;
```
Comment (banmoy) [inline starrocks-stream-load-sdk/src/main/java/com/starrocks/data/load/stream/v2/TransactionTableRegion.java:343]: **[Suggestion] Extract spin-lock pattern into utility**

The spin-lock pattern with `MAX_SPIN_ATTEMPTS`, `Thread.yield()`, and `LockSupport.parkNanos()` is duplicated in `write()`, `switchChunkForCommit()`, and `flush()`. Consider extracting into a utility method or a small `SpinLock` class. Alternatively, `ReentrantLock.tryLock()` would provide the same performance with better debuggability (thread dump shows lock owner).

**Review by banmoy — COMMENTED** [review]
## Code Review Summary

**Architecture**: Clean separation between PartitionCommitTracker (state machine), SharedTransactionCoordinator (label lifecycle), and DefaultStreamLoadManager (orchestration). Well-designed.

**Test coverage**: Comprehensive — 8 PartitionCommitTracker unit tests, 17 StreamLoadManagerMultiTableTest unit tests, 11 integration tests covering edge cases.

**Backward compatibility**: Non-multi-table paths carefully guarded. Good.

### Issues Found

| Severity | Count | Summary |
|---|---|---|
| Critical | 3 | `setLabel()` behavior change, misleading Javadoc, double savepoint reset |
| Important | 5 | Unconditional headers, no commitInFlight timeout, recycle ordering, partitionRegions leak, missing V2 validation |
| Suggestion | 1 | Extract spin-lock pattern |

See inline comments for details.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by banmoy** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Review by banmoy — APPROVED** [review]

**Comment by jaogoy** [comment]
So many code lines.
